From 766366703b43100f0d06d8c71eeabc229119c9da Mon Sep 17 00:00:00 2001
From: Subbaraya Sundeep <sbhatta@marvell.com>
Date: Thu, 6 Jun 2019 12:27:19 +0530
Subject: [PATCH 029/138] octeontx2-pf: Cleanup properly during driver remove

commit fdb7543bc0cdec7b680ab82ba7e017c4a64eff35 from
git@git.assembla.com:cavium/WindRiver.linux.git

When unbinding a PF driver sriov needs to be disabled
before disabling PF interrupts. Otherwise all the VF cleanup
messages never get response from AF since PF interrupt is
disabled. Also VF ntuple filters installed by a PF needs to
be deleted before VF removal otherwise already running
traffic may hit VF ntuple filter rules for removed VFs.

Change-Id: I1e7cefde646dc88ef5b14387e00a709b7f2d8876
Signed-off-by: Subbaraya Sundeep <sbhatta@marvell.com>
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c | 11 +++++------
 1 file changed, 5 insertions(+), 6 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
index 813072b34f66..7952ab642d56 100644
--- a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
+++ b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
@@ -2413,6 +2413,7 @@ static int otx2_sriov_disable(struct pci_dev *pdev)
 	if (!pci_num_vf(pdev))
 		return 0;
 
+	otx2_delete_vf_ethtool_flows(pf);
 	pci_disable_sriov(pdev);
 
 	for (i = 0; i < pci_num_vf(pdev); i++)
@@ -2421,7 +2422,6 @@ static int otx2_sriov_disable(struct pci_dev *pdev)
 
 	otx2_disable_flr_me_intr(pf);
 	otx2_flr_wq_destroy(pf);
-	otx2_delete_vf_ethtool_flows(pf);
 	otx2_disable_pfvf_mbox_intr(pf);
 	otx2_pfvf_mbox_destroy(pf);
 
@@ -2455,15 +2455,14 @@ static void otx2_remove(struct pci_dev *pdev)
 	otx2_cgx_config_linkevents(pf, false);
 
 	unregister_netdev(netdev);
-	otx2_destroy_ethtool_flows(pf);
-	otx2_ptp_destroy(pf);
 
-	otx2_disable_mbox_intr(pf);
+	otx2_sriov_disable(pf->pdev);
+	otx2_ptp_destroy(pf);
+	otx2_destroy_ethtool_flows(pf);
 
 	otx2_detach_resources(&pf->mbox);
+	otx2_disable_mbox_intr(pf);
 	otx2_pfaf_mbox_destroy(pf);
-	otx2_sriov_disable(pf->pdev);
-
 	pci_free_irq_vectors(pf->pdev);
 	pci_set_drvdata(pdev, NULL);
 	free_netdev(netdev);
-- 
2.17.1

